### PR TITLE
Add support for exporting colors from variables

### PR DESCRIPTION
**Pull request by alexey1312**
- Added `ColorsVariablesLoader`
- Updated doc

[Scheme](https://www.figma.com/file/sXTVIxS0WMGuMSx9WloJk3/Syncing-design-systems-using-Variables-REST-API-(Community)?type=whiteboard&node-id=0-1&t=Gp42pIzJsf6WEMdW-0)

CleanShot 2024-04-04 at 08 12 40@2x

From the documentation:

1. tokensCollectionName - the name of the variable collection
2. lightModeName - the name of the color variable column for the light theme
3. darkModeName - the name of the color variable column for the dark theme
4. lightHCModeName - the name of the color variable column for the light theme with high contrast
5. darkHCModeName - the name of the color variable column for the dark theme with high contrast
6. A variable that has a local value
7. A variable that refers to another variable in a different file

1. primitivesModeName - the name of the variable column, if the value in [CONFIG.md](CONFIG.md) is not specified, the default value will be used
2. A variable that has a local value

fyi:
https://github.com/RedMadRobot/figma-export/issues/217

### CODE REVIEW

**Review by subdan — CHANGES_REQUESTED** [review]

File: README.md [hunk 1]
```diff
@@ -490,11 +496,46 @@ Example
 | <img src="images/dark.png" width="352" />  | <img src="images/dark_c.png" width="200" />  |
 | <img src="images/light.png" width="352" /> | <img src="images/light_c.png" width="200" /> |
 
+### For variables
+
+For `figma-export colors`
+
+**Important, the [API](https://www.figma.com/developers/api#variables) for working with color variables in Figma is still in `Beta` stage, so something may break at any time.**
+
+**Important, in [CONFIG.md](CONFIG.md) use either colors or variablesColors.**
```
[the remark this answers is not in the document]
Comment (alexey1312) [reply README.md]: done ✅

File: README.md [hunk 2]
```diff
@@ -490,11 +496,46 @@ Example
...
+
+With the introduction of color variables in Figma, you can use color variables instead of color styles. Color variables can be used in figma-export, for this in [CONFIG.md](CONFIG.md) you need to use the `variablesColors` variable instead of `colors`.
```
[the remark this answers is not in the document]
Comment (alexey1312) [reply README.md]: done ✅

File: README.md [hunk 3]
```diff
@@ -490,11 +496,46 @@ Example
...
+5. darkHCModeName - the name of the color variable column for the dark theme with high contrast
+6. A variable that has a local value
+7. A variable that refers to another variable in a different file
+
+<img src="images/figma_colors_primitives.png" width="352" />
+
+1. primitivesModeName - the name of the variable column, if the value in [CONFIG.md](CONFIG.md) is not specified, the default value will be used
+2. A variable that has a local value
+
+See [CONFIG.md](CONFIG.md) for more information in the variablesColors section.
```
[the remark this answers is not in the document]
Comment (alexey1312) [reply README.md]: done ✅

File: README.md [hunk 4]
```diff
@@ -490,11 +496,46 @@ Example
...
+2. lightModeName - the name of the color variable column for the light theme
+3. darkModeName - the name of the color variable column for the dark theme
+4. lightHCModeName - the name of the color variable column for the light theme with high contrast
+5. darkHCModeName - the name of the color variable column for the dark theme with high contrast
+6. A variable that has a local value
+7. A variable that refers to another variable in a different file
+
+<img src="images/figma_colors_primitives.png" width="352" />
+
+1. primitivesModeName - the name of the variable column, if the value in [CONFIG.md](CONFIG.md) is not specified, the default value will be used
```
[the remark this answers is not in the document]
Comment (alexey1312) [reply README.md]: done ✅

File: Sources/FigmaExport/Subcommands/ExportColors.swift [hunk 5]
```diff
@@ -25,20 +25,56 @@ extension FigmaExportCommand {
...
+                loader = ColorsVariablesLoader(
+                    client: client,
+                    figmaParams: figmaParams,
+                    variableParams: variableParams,
+                    filter: filter
+                )
+                colors = try loader.load()
+
+                nameValidateRegexp = options.params.common?.variablesColors?.nameValidateRegexp
+                nameReplaceRegexp = options.params.common?.variablesColors?.nameReplaceRegexp
```
[the remark this answers is not in the document]
Comment (alexey1312) [reply Sources/FigmaExport/Subcommands/ExportColors.swift]: done ✅

File: Sources/FigmaExport/Subcommands/ExportColors.swift [hunk 6]
```diff
@@ -25,20 +25,56 @@ extension FigmaExportCommand {
...
-            let loader = ColorsLoader(client: client, figmaParams: options.params.figma, colorParams: options.params.common?.colors)
-            let colors = try loader.load(filter: filter)
+            let figmaParams = options.params.figma
+            var colors: ColorsLoaderOutput?
+            var loader: ColorsLoaderProtocol
+            var nameValidateRegexp: String?
+            var nameReplaceRegexp: String?
+
+            if let colorParams = commonParams?.colors {
+                loader = ColorsLoader(
```
[the remark this answers is not in the document]
Comment (alexey1312) [reply Sources/FigmaExport/Subcommands/ExportColors.swift]: done ✅

File: Sources/FigmaAPI/Model/Variables.swift [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    public var resolvedType: ResolvedType
+    public var valuesByMode: [String: ValuesByMode]
+    public var remote: Bool
+    public var description: String
+    public var hiddenFromPublishing: Bool
+}
+
+public struct VariablesMeta: Decodable {
+    public var variableCollections: [String: VariableCollectionId]
+    public var variables: [String: VariableID]
```
[the remark this answers is not in the document]
Comment (alexey1312) [reply Sources/FigmaAPI/Model/Variables.swift]: done ✅

File: Sources/FigmaAPI/Model/Variables.swift [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        if let variableAlias = try? VariableAlias(from: decoder) {
+            self = .variableAlias(variableAlias)
+        } else if let color = try? PaintColor(from: decoder) {
+            self = .color(color)
+        } else if let string = try? String(from: decoder) {
+            self = .string(string)
+        } else if let number = try? Double(from: decoder) {
+            self = .number(number)
+        } else if let boolean = try? Bool(from: decoder) {
+            self = .boolean(boolean)
```
[the remark this answers is not in the document]
Comment (alexey1312) [reply Sources/FigmaAPI/Model/Variables.swift:44]: done ✅

File: Sources/FigmaAPI/Model/Variables.swift [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+public struct VariableCollectionId: Decodable {
+    public var defaultModeId: String
+    public var id: String
+    public var name: String
+    public var remote: Bool
+    public var modes: [Mode]
+    public var key: String
+    public var hiddenFromPublishing: Bool
```
[the remark this answers is not in the document]
Comment (alexey1312) [reply Sources/FigmaAPI/Model/Variables.swift]: done ✅

File: Sources/FigmaAPI/Model/Variables.swift [hunk 10]
```diff
@@ -0,0 +1,82 @@
+public struct Mode: Decodable {
+    public var modeId: String
+    public var name: String
+}
+
+public struct VariableCollectionId: Decodable {
```
[the remark this answers is not in the document]
Comment (alexey1312) [reply Sources/FigmaAPI/Model/Variables.swift]: done ✅

File: Sources/FigmaExport/Subcommands/ExportColors.swift [hunk 11]
```diff
@@ -25,20 +25,56 @@ extension FigmaExportCommand {
...
+                    client: client,
+                    figmaParams: figmaParams,
+                    colorParams: colorParams,
+                    filter: filter
+                )
+                colors = try loader.load()
+                
+                nameValidateRegexp = options.params.common?.colors?.nameValidateRegexp
+                nameReplaceRegexp = options.params.common?.colors?.nameReplaceRegexp
```
[the remark this answers is not in the document]
Comment (alexey1312) [reply Sources/FigmaExport/Subcommands/ExportColors.swift]: done ✅

**Pull request merged**